src/runtime/storageNG/direct-store.ts

```diff
@@ -0,0 +1,241 @@
+    }
+  }
+
+  private async processModelChange(modelChange: CRDTChange<T>, otherChange: CRDTChange<T>, version: number, fromDriver: boolean) {
+    this.deliverCallbacks(modelChange);
+    await this.updateStateAndAct(this.noDriverSideChanges(modelChange, otherChange, fromDriver), version, fromDriver);
+  }
+  // This function implements a state machine that controls when data is sent to the driver.
+  // You can see the state machine in all its glory at the following URL:
+  //
+  // https://github.com/PolymerLabs/arcs/wiki/Store-object-State-Machine
+  //
+  private async updateStateAndAct(noDriverSideChanges: boolean, version: number, messageFromDriver: boolean) {
+    // Don't send to the driver if we're already in sync and there are no driver-side changes.
+    if (noDriverSideChanges) {
+      // Need to record the driver version so that we can continue to send.
+      this.setState(DirectStoreState.Idle);
+      this.version = version;
+      return;
+    }
+    switch (this.state) {
+      case DirectStoreState.AwaitingDriverModel:
+        if (!messageFromDriver) {
+          return;
+        }
+      /* falls through */
+      case DirectStoreState.Idle:
+        // This loop implements sending -> AwaitingResponse -> AwaitingResponseDirty -> sending.
+        // Breakouts happen if:
+        //  (1) a response arrives while still AwaitingResponse. This returns the store to Idle.
+        //  (2) a negative response arrives. This means we're now waiting for driver models
+        //      (AwaitingDriverModel). Note that in this case we are likely to end up back in
+        //      this loop when a driver model arrives.
+        while (true) {
+          this.setState(DirectStoreState.AwaitingResponse);
+          // Work around a typescript compiler bug. Apparently typescript won't guarantee that
+          // a Map key you've just set will exist, but is happy to assure you that a private
+          // member variable couldn't possibly change in any function outside the local scope
+          // when within a switch statement. 
+          this.state = DirectStoreState.AwaitingResponse;
+          version += 1;
+          this.version = version;
+          const response = await this.driver.send(this.localModel.getData(), version);
+          if (response) {
+            if (this.state === DirectStoreState.AwaitingResponse) {
+              this.setState(DirectStoreState.Idle);
+              this.applyPendingDriverModels();
+              break;
+            }
+            if (this.state !== DirectStoreState.AwaitingResponseDirty) {
+              // This shouldn't be possible as only a 'nack' should put us into
+              // AwaitingDriverModel, and only the above code should put us back
+              // into Idle.
+              throw new Error('reached impossible state in store state machine');
+            }
+            // fallthrough to re-execute the loop.
```

Comment:
Instead of comment, just say `continue;`

Reply:
I think that would be less clear.

Reply:
`continue` says in code exactly what your comment says, except the compiler understands it too.